p2p/protocol/holepunch/svc.go

```diff
@@ -106,7 +106,8 @@ func (s *Service) watchForPublicAddr() {
 	t := time.NewTimer(duration)
 	defer t.Stop()
 	for {
-		if containsPublicAddr(s.ids.OwnObservedAddrs()) {
+		// Use both host and observed to enable hole punching for undialable public ips that might not be observed
+		if containsPublicAddr(s.ids.OwnObservedAddrs()) || containsPublicAddr(s.host.Addrs()) {
```

Comment:
This won't work with hosts that have autorelay enabled. https://github.com/libp2p/go-libp2p/blob/master/p2p/host/autorelay/relay_finder.go#L728
Autorelay wraps host.Addrs and filters out all public host addresses and adds relayed addresses. 

Instead, we can do, `host.Network().ListenAddresses()` and `host.Network().InterfaceListenAddresses()`

Reply:
What's the difference between `host.Network().ListenAddresses()` and `host.Network().InterfaceListenAddresses()`?

Reply:
The transport listen on `/ip4/0.0.0.0/tcp/12345` then `InterfaceListenAddresses()` use https://pkg.go.dev/net#Interfaces to explode this to:
- `/ip4/127.0.0.1/tcp/12345`
- `/ip4/192.168.1.123/tcp/12345`
- `/ip4/10.13.37.42/tcp/12345`

(numbers made up for example sake, varies based on the machine config)

Reply:
Given that, we probably only want to rely on `host.Network().InterfaceListenAddresses()` here to get the actual IPs of the interfaces. 

Or is there a case where we want to also check `host.Network().ListenAddresses()`?